xdsl/dialects/arith.py

```diff
@@ -395,19 +425,27 @@ def infer_overflow_type(input_type: Attribute) -> Attribute:
         )
 
 
-class MuliHasCanonicalizationPatterns(HasCanonicalizationPatternsTrait):
-    @classmethod
-    def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
-        from xdsl.transforms.canonicalization_patterns import arith
-
-        return (arith.MuliIdentityRight(), arith.MuliConstantProp())
-
-
 @irdl_op_definition
 class MuliOp(SignlessIntegerBinaryOperationWithOverflow):
     name = "arith.muli"
 
-    traits = traits_def(Pure(), MuliHasCanonicalizationPatterns())
+    traits = traits_def(
+        Pure(),
+        Commutative(),
+        SignlessIntegerBinaryOperationHasCanonicalizationPatternsTrait(),
+    )
+
+    @staticmethod
+    def py_operation(lhs: int, rhs: int) -> int | None:
+        return lhs * rhs
+
+    @staticmethod
+    def is_right_unit(attr: AnyIntegerAttr) -> bool:
+        return attr == IntegerAttr(1, attr.type)
```

Comment:
Why is this one taking the type into account, but the one in Addi just checks the value?

Reply:
The one here is just to work around that the unit for `i1` multiplication is `-1` and not `1`. This doesn't matter in the 0 case which `Addi` has

Reply:
Thinking about it I'm not 100% sure this will work for `si1`

Reply:
@superlopuh is there a better way to do this check?

Reply:
this is what we seem to be doing in other places too

Reply:
I can't think of one